PreStop Event Feature

maysunfaisal opened the pull request:
Signed-off-by: Maysun J Faisal <maysun.j.faisal@ibm.com>

**What type of PR is this?**
/kind feature

**What does does this PR do / why we need it**:
- Executes `preStop` events, if present, before deleting the component
- Minor `odo delete` label/heading changes to reflect addition of preStop feature
- Add show flag to postStart event execution
- Add show flag to delete command on experimental
- Validate devfile events before using them
- Unit & Integration tests

**Which issue(s) this PR fixes**:

Fixes #3566
Fixes #3632
Fixes #3655

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

**How to test changes / Special notes to the reviewer**:
- Add in preStop events to a devfile or you may try with this sample:
```
schemaVersion: 2.0.0
metadata:
  name: nodejs
projects:
  - name: nodejs-starter
    git:
      location: "https://github.com/odo-devfiles/nodejs-ex.git"
components:
  - container:
      name: runtime
      image: quay.io/eclipse/che-nodejs10-ubi:nightly
      memoryLimit: 1024Mi
      endpoints:
        - name: "3000/tcp"
          configuration:
            protocol: tcp
            scheme: http
          targetPort: 3000 
      mountSources: true
  - container:
      name: "tools"
      image: quay.io/eclipse/che-nodejs10-ubi:nightly
      mountSources: true
      memoryLimit: 1024Mi
commands:
  - exec:
      id: myPostStart
      commandLine: echo I am a PostStart
      component: tools
      workingDir: /
  - exec:
      id: secondPostStart
      commandLine: echo I am also a PostStart
      component: runtime
      workingDir: /
  - exec:
      id: devbuild
      component: runtime
      commandLine: npm install
      workingDir: ${CHE_PROJECTS_ROOT}/nodejs-starter
      group:
        kind: build
        isDefault: false
  - exec:
      id: build
      component: runtime
      commandLine: npm install
      workingDir: ${CHE_PROJECTS_ROOT}/nodejs-starter
      group:
        kind: build
  - exec:
      id: devrun
      component: runtime
      commandLine: npm start
      workingDir: ${CHE_PROJECTS_ROOT}/nodejs-starter
      group:
        kind: run
        isDefault: true
  - exec:
      id: run
      component: runtime
      commandLine: npm start
      workingDir: ${CHE_PROJECTS_ROOT}/nodejs-starter
      group:
        kind: run
  - composite:
         id: mycompevent
         label: Build and Mkdir
         commands:
           - myPostStart
           - secondPostStart
         parallel: true
         group:
          kind: build
          isDefault: true
events:
  preStart:
    - "myPostStart" 
    - "secondPostStarssst"
  postStart:
    - "mycompevent" 
    - "secondPostStart"
  preStop:
    - "mycompevent" 
    - "secondPostStart"
  postStop:
    - "myPostStart" 
    - "secondPostStart"
```

should output:
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo delete -f

Gathering information for component madridnode
 ✓  Checking status for component madridnode [132ms]

Executing preStop event commands for component madridnode
 •  Executing secondpoststart command "echo I am also a PostStart"  ...
 •  Executing mypoststart command "echo I am a PostStart"  ...
 ✓  Executing secondpoststart command "echo I am also a PostStart" [1s]
 ✓  Executing mypoststart command "echo I am a PostStart" [1s]
 ✓  Executing secondpoststart command "echo I am also a PostStart" [941ms]

Deleting devfile component madridnode
 ✓  Deleting Kubernetes resources for component madridnode [129ms]
 ✓  Successfully deleted component
```

mik-dass commented on pkg/devfile/validate/events.go:
```diff
@@ -0,0 +1,73 @@
+	return nil
+}
+
+func isEventValid(eventNames []string, eventType string, commands []common.DevfileCommand) error {
```
comments for the function

maysunfaisal replied:
updated

mik-dass commented on pkg/devfile/validate/events_test.go:
```diff
@@ -0,0 +1,157 @@
+			errorShouldContain: nil,
+		},
+		{
+			name: "Case 1: Valid events",
```
the name of the case is the same as the one before

maysunfaisal replied:
updated

mik-dass commented on pkg/testingutil/pods.go:
```diff
@@ -0,0 +1,22 @@
+)
+
+// CreateFakePod creates a fake pod with the given pod name and component name
+func CreateFakePod(componentName, podName string) *corev1.Pod {
```
this should be moved here https://github.com/openshift/odo/tree/master/pkg/kclient/fake

maysunfaisal replied:
hmm do we want it there? there are loads of Fake resource creation under pkg/testingutil like this https://github.com/openshift/odo/blob/master/pkg/testingutil/deployments.go

mik-dass replied:
Ok let's keep it in testingutil for now.

mik-dass commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -569,27 +568,46 @@ func getFirstContainerWithSourceVolume(containers []corev1.Container) (string, s
 
-	componentExists, err := utils.ComponentExists(a.Client, a.ComponentName)
+	log.Infof("\nGathering information for component %s", a.ComponentName)
+	podSpinner := log.Spinnerf("Retrieving pod for component %s", a.ComponentName)
```
We shouldn't display terms like pod in the UI. This should be something simpler like `Checking component status` or `Retrieving component info`

maysunfaisal replied:
I am going to say `Retrieving info for component %s`

maysunfaisal replied:
updated

mik-dass commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -163,14 +163,11 @@ func (a Adapter) Push(parameters common.PushParameters) (err error) {
 	postStartEvents := a.Devfile.Data.GetEvents().PostStart
 	if !componentExists && len(postStartEvents) > 0 {
-		// log only when there are post start events present in devfile
-		log.Infof("\nExecuting postStart event commands for component %s", a.ComponentName)
```
Any reason why the log statement was removed?

mik-dass commented on pkg/devfile/validate/events_test.go:
```diff
@@ -0,0 +1,157 @@
+		errorShouldContain []string
+	}{
+		{
+			name: "Case 1: Valid events",
```
please add cases with postStop and preStart events.

mik-dass commented:
> How to test changes / Special notes to the reviewer:

Please add some steps to test the changes in the PR.

maysunfaisal commented:
> > How to test changes / Special notes to the reviewer:
> 
> Please add some steps to test the changes in the PR.

@mik-dass updated the PR OP

mik-dass commented:
```
[mrinaldas@localhost project]$ odo delete -f
I0729 11:45:13.425475   13759 cert_rotation.go:137] Starting client certificate rotation controller

Gathering information for component nodejs
 ✓  Checking status for component nodejs [5ms]

Executing preStop event commands for component nodejs
 ✗  
Unable to exec command [/bin/sh -c cd /project && echo i am something]: 

 ✗  Executing some_exec command "echo i am something" [6ms]
 ✗  error occurred while deleting component, cause: unable to execute devfile command some_exec: unable to execute the build command: error while streaming command: container runtime1 is not valid for pod nodejs-55b7c7747b-xrlvq
```
I think we should also validate that the command, referred to by the event, has a valid component. This error message seems pretty confusing to be TBH.

maysunfaisal commented:
> I think we should also validate that the command, referred to by the event, has a valid component. This error message seems pretty confusing to be TBH.

@mik-dass good catch. I have updated the PR with those validations and also updated unit & integration tests for it. 👍 

I also found a new bug for composite commands's child commands and opened an issue here https://github.com/openshift/odo/issues/3656. This issue is for composite commands is also seen with odo push, so that should be addressed separately.

maysunfaisal commented:
/retest

mik-dass commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -569,27 +568,46 @@ func getFirstContainerWithSourceVolume(containers []corev1.Container) (string, s
+		}
 	}
 
+	log.Infof("\nDeleting devfile component %s", a.ComponentName)
```
Can we simply say `Deleting component` instead? I feel it sounds better with `Checking status for component` and `Gathering information for component`.

maysunfaisal replied:
updated, I also did a few more modifications like removing the component name from the spinner. I think it is redundant to always say component name in spinner if the header has that info already. So this is how the updated header and spinner looks:
```
Maysuns-MacBook-Pro:nodejs-ex maysun$ odo delete -f

Gathering information for component madridnode
 ✓  Checking status for component [125ms]

Executing preStop event commands for component madridnode
 •  Executing secondpoststart command "echo I am also a PostStart"  ...
 •  Executing mypoststart command "echo I am a PostStart"  ...
 ✓  Executing mypoststart command "echo I am a PostStart" [924ms]
 ✓  Executing secondpoststart command "echo I am also a PostStart" [962ms]
 ✓  Executing secondpoststart command "echo I am also a PostStart" [945ms]

Deleting component madridnode
 ✓  Deleting Kubernetes resources for component [132ms]
 ✓  Successfully deleted component
```

maysunfaisal commented:
/retest

mik-dass commented:
Work for me :+1: but the PR needs a rebase

adisky commented on pkg/kclient/errors.go:
```diff
@@ -0,0 +1,12 @@
+import "fmt"
+
+// NoPodFoundError returns an error if no pod is found with the selector
+type NoPodFoundError struct {
```
s/NoPodFoundError/PodNotFoundError?

adisky commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -569,27 +568,46 @@ func getFirstContainerWithSourceVolume(containers []corev1.Container) (string, s
+		podSpinner.End(false)
 		log.Warningf("%v", err)
 		return nil
+	} else if e, ok := err.(*kclient.NoPodFoundError); ok {
```
wouldn't this error(kclient.NoPodFoundError) be returned by kubernetes also? do we need to define a separate one?

maysunfaisal replied:
The kclient API uses a list API, so in case of 0, it returns an empty array. We had to change the kclient API being used with this PR from `utils.ComponentExists()` to `a.Client.GetPodUsingComponentName()`

here is what it does in kclient:
```
pods, err := c.KubeClient.CoreV1().Pods(c.Namespace).List(metav1.ListOptions{
		LabelSelector: selector,
	})
	if err != nil {
		// Dont wrap error since we want to know if its a forbidden error
		// if we wrap, we lose the err status reason and callers of this api rely on it
		return nil, err
	}
	numPods := len(pods.Items)
	if numPods == 0 {
		return nil, &NoPodFoundError{Selector: selector}
	} else if numPods > 1 {
		return nil, fmt.Errorf("multiple Pods exist for the selector: %v. Only one must be present", selector)
	}
```

adisky commented on pkg/kclient/errors.go:
```diff
@@ -0,0 +1,12 @@
+
+// NoPodFoundError returns an error if no pod is found with the selector
+type NoPodFoundError struct {
+	Selector string
```
s/NoPodFoundError/PodNotFoundError?

maysunfaisal replied:
will update it, thx

adisky commented on pkg/devfile/validate/events.go:
```diff
@@ -0,0 +1,96 @@
+				isEventPresent = true
+
+				// Check if the devfile command is valid
+				err := adapterCommon.ValidateCommand(data, command)
```
commands should be validated separately(not saying to resolve in this PR)

maysunfaisal replied:
How would you propose command to be validated that does not involve `odo push`? Here I am reusing the func we use to validate commands for `odo push`

adisky replied:
all devfile commands should be validated just after parsing, like we are doing for `events`,  if all commands validated before only, no need to validate them here.

adisky replied:
https://github.com/openshift/odo/pull/3631/files#diff-838aff5d94d44c245e0db507c96219d9R24

maysunfaisal replied:
Agreed, opened an issue to refactor it https://github.com/openshift/odo/issues/3703

johnmcollier commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -569,27 +568,46 @@ func getFirstContainerWithSourceVolume(containers []corev1.Container) (string, s
+	// if there are preStop events, execute them before deleting the deployment
+	preStopEvents := a.Devfile.Data.GetEvents().PreStop
+	if len(preStopEvents) > 0 {
+		if pod.Status.Phase != corev1.PodRunning {
```
Rather than manually checking if the phase is `PodRunning`, can we just call `WaitAndGetPod` when we initally retrieve the pod and set the phase to wait for too `PodRunning`?

maysunfaisal replied:
It was a conscious decision to do this. A while ago, @mik-dass reviwed Stephanie's PR that I was also reviewing and had this same conversation https://github.com/openshift/odo/pull/3400#pullrequestreview-443650340

So I did this to be consistent and since @mik-dass is also reviewing this PR. Also, since we need the pod name, we already got the pod info in the beginning of `Delete()`, and agree with @mik-dass on the other PR that there is no need to do an API call again.

johnmcollier replied:
Could we wrap it into a function then when we get the pod? We retrieve the pod and verify `pod.Status.Phase` with the requested phase?

maysunfaisal replied:
Don't really want to introduce another helper/wrapper func, there are already so many in pods.go https://github.com/openshift/odo/blob/master/pkg/kclient/pods.go#L154 and we're not being uniform in its usage everywhere. What is the wrapper going to achieve here, just return the pod name and status?

johnmcollier replied:
That's fair.

My concern is that if we're not going to be using `WaitAndGetPod`, there's going to be loads of duplicated code in odo that pretty much does this (gets the pod, and then retrieves / checks the status).

johnmcollier replied:
This case may not fit well for a helper function, since we're not checking the pod status until a bit after we get the pod, but in general, if we're frequently getting a pod and immediately checking its status, then that should be rolled up into a helper function.

maysunfaisal replied:
Hmm i see your point, but it will pretty much be the same thing:

currently:
```
pod, err := a.Client.GetPodUsingComponentName(a.ComponentName)
	if err != nil {
		return fmt.Errorf("error occurred while getting the pod: %w", err)
	}
	if pod.Status.Phase != corev1.PodRunning {
		return fmt.Errorf("pod for component %s is not running", a.ComponentName)
	}
```

Some callers are using `pod.Name` and all are checking for the status. So will the wrapper just be returning the pod{}, err, isPodRunning ?

maysunfaisal commented:
@mik-dass @adisky rebased and updated PR, pls help review/lgtm it, thx ~

johnmcollier commented on pkg/devfile/adapters/kubernetes/component/adapter.go:
```diff
@@ -569,27 +568,46 @@ func getFirstContainerWithSourceVolume(containers []corev1.Container) (string, s
+func (a Adapter) Delete(labels map[string]string, show bool) error {
 
-	componentExists, err := utils.ComponentExists(a.Client, a.ComponentName)
+	log.Infof("\nGathering information for component %s", a.ComponentName)
```
Do we need both the `nGathering information for component` and `log.Spinner("Checking status for component")`?

Think we'd just want the spinner?

johnmcollier commented on pkg/odo/cli/component/delete.go:
```diff
@@ -48,12 +48,13 @@ type DeleteOptions struct {
 // NewDeleteOptions returns new instance of DeleteOptions
 func NewDeleteOptions() *DeleteOptions {
-	return &DeleteOptions{false, false, false, false, "", false, &ComponentOptions{}, "", "", nil}
+	return &DeleteOptions{false, false, false, false, "", false, &ComponentOptions{}, "", "", false, nil}
```
Since you're adding another parameter to the struct, could we reformat this to make it a little more readable and to see which parameters are getting set?
```
return &DeleteOptions{
    componentForceDeleteFlag: false, 
    componentDeleteAllFlag: false,
    componentDeleteWaitFlag: false,
    componentDeleteS2iFlag: false, 
    componentContext: "", 
    isCmpExists: false, 
    ...
}

maysunfaisal replied:
good point, updated

maysunfaisal replied:
@johnmcollier pls help review the PR, thx

johnmcollier replied:
👍

maysunfaisal commented:
/retest

adisky commented:
/lgtm

kadel commented:
/approve
